top-n (#15)
ververica/flink-sql-cookbook: recipes/11/11_top_n.md

# 11 Continuous Top-N

:bulb: This example will show to to calculate a continuous Top-N using an `OVER` window.
A typical use case is tracking the most popular rows of some set.

The source table (`spells_cast`) is backed by the [`faker` connector](https://github.com/knaufk/flink-faker), which continuously generates rows in memory based on Java Faker expressions.

> The Ministry of Magic tracks every spell a wizard casts throughout Great Britain and wants to know every wizard's top 2 all-time favorite spells. 

Reviewer: ```suggestion
The Ministry of Magic tracks every spell a wizard casts throughout Great Britain and wants to know every wizard's Top 2 all-time favorite spells. 
```

> Flink SQL can calculate continuous [aggregations](../03/03_group_by.md), so if we know

Reviewer: ```suggestion
Flink SQL can be used to calculate continuous [aggregations](../03/03_group_by.md), so if we know
```

each spell a wizard has cast, we can maintain a continuous total of how many times they have cast that spell. 

> ```sql

Reviewer: Maybe drop the DDL here and use just the query?

CREATE TABLE spells_cast (
    wizard STRING,
    spell  STRING
) WITH (...);

SELECT wizard, spell, COUNT(*) AS times_cast
FROM spells_cast
GROUP BY wizard;
```

This result can be used in an `OVER` window to calculate a [Top-N](https://docs.ververica.com/user_guide/sql_development/queries.html#top-n).

> The columns `wizard` partitions the rows, and they are then ordered based on `times_cast DESC`. 

Reviewer: ```suggestion
The rows are partitioned using the `wizard` column, and are then ordered based on the count of spell casts (`times_cast DESC`). 
```
